### PR TITLE
feat: Provide chainId prop for Asset List

### PR DESCRIPTION
**Pull request by darkwing**
## **Description**

This PR sets shows the network badge specific to the asset, not the globally selected chain ID.  Of course for now, that chainId is shimmed to be the current chainId until we have the Accounts API providing those assets in correct form

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27724?quickstart=1)

## ** ToDo **
- [x] Ensure that the AssetPicker in Send and Swaps looks correct
- [x] Do the same for the Assets Detail page

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to token list and check the network badge 
2. Go to asset picker and check the network badge
3. Go to asset details and check the network badge

### **Before**

SCR-20241009-irdt
SCR-20241009-irgl

### **After**

Should be the same
Screenshot 2024-10-17 at 20 37 16

Screenshot 2024-10-17 at 20 37 47

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/app/assets/token-cell/token-cell.tsx [hunk 1]
```diff
@@ -32,11 +34,12 @@ export default function TokenCell({
...
-    minimumSignificantDigits: 1,
-    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-    // @ts-ignore
-  }).format(string.toString());
+  const primary =
+    new Intl.NumberFormat(locale, {
+      minimumSignificantDigits: 1,
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+    }).format(string.toString()) || '0';
```
Comment (gambinish) [inline ui/components/app/assets/token-cell/token-cell.tsx:42]: Is this `|| '0'` needed for this change?

File: ui/components/app/assets/token-list/token-list.tsx [hunk 2]
```diff
@@ -85,7 +87,7 @@ export default function TokenList({
         return (
           <TokenCell
             key={`${tokenData.symbol}-${tokenData.address}`}
-            {...tokenData}
+            {...{ ...tokenData, chainId }}
```
Comment (gambinish) [inline ui/components/app/assets/token-list/token-list.tsx:90]: Rather than double spreading, I think passing in `chainId` as a dedicated prop is a bit cleaner:
```suggestion
            {...tokenData}
            chainId={chainId}
```

File: ui/components/multichain/nft-item/nft-item.js [hunk 3]
```diff
@@ -16,27 +16,35 @@ import {
...
+  const { image, imageOriginal, name, tokenId } = nft;
+
+  const ipfsGateway = useSelector(getIpfsGateway);
+  const currentChain = useSelector(getCurrentNetwork);
+  const nftImageURL = useGetAssetImageUrl(imageOriginal ?? image, ipfsGateway);
+
+  const nftImageAlt = getNftImageAlt(nft);
+  const nftSrcUrl = imageOriginal ?? image;
+  const isIpfsURL = nftSrcUrl?.startsWith('ipfs:');
+  const isImageHosted = image?.startsWith('https:');
```
Comment (gambinish) [inline ui/components/multichain/nft-item/nft-item.js:41]: No need for this change, but I typically prefer to use the [URL.protocol](https://developer.mozilla.org/en-US/docs/Web/API/URL/protocol) property for this sort of thing.

File: ui/components/app/assets/token-cell/token-cell.tsx (hunk 1 above)
Comment (salimtb) [reply ui/components/app/assets/token-cell/token-cell.tsx:42]: i think so yes, the `format(string.toString()) || '0'` ensures that string is properly formatted and falls back to '0' if the value is `null` or `undefined`.

File: ui/components/app/assets/token-list/token-list.tsx (hunk 2 above)
Comment (salimtb) [reply ui/components/app/assets/token-list/token-list.tsx]: fixed [0912127](https://github.com/MetaMask/metamask-extension/pull/27724/commits/0912127050ba6e647f334f6be4588634e214d47d)

File: ui/components/multichain/nft-item/nft-item.js (hunk 3 above)
Comment (salimtb) [reply ui/components/multichain/nft-item/nft-item.js]: fixed [0912127](https://github.com/MetaMask/metamask-extension/pull/27724/commits/0912127050ba6e647f334f6be4588634e214d47d)

Comment (salimtb) [reply ui/components/multichain/nft-item/nft-item.js]: Good point! However, using URL.protocol can throw an error if the string is not a valid URL. That's why I prefer to handle it safely by either validating the input or wrapping it in a try-catch to avoid potential issues with malformed URLs.

**Review by gambinish — DISMISSED** [review]

**Review by NidhiKJha — DISMISSED** [review]

File: ui/components/app/assets/token-list/token-list.tsx [hunk 4]
```diff
@@ -30,6 +31,7 @@ export default function TokenList({
   nativeToken,
 }: TokenListProps) {
   const t = useI18nContext();
+  const chainId = useSelector(getMultichainCurrentChainId);
   const { tokenSortConfig } = useSelector(getPreferences);
```
Comment (sahar-fehri) [inline ui/components/app/assets/token-list/token-list.tsx:35]: Im seeing that to retrieve the chainId we sometimes use 
`const chainId = useSelector(getMultichainCurrentChainId);`
or
`const chainId = useSelector(getCurrentChainId);`

Is this something we can consolidate?

Comment (darkwing) [reply ui/components/app/assets/token-list/token-list.tsx:35]: Good shout @sahar-fehri ; I've updated to use `getMultichainCurrentChainId`

**Review by Joj501 — APPROVED** [review]
Change

**Review by sahar-fehri — APPROVED** [review]

**Review by vinnyhoward — APPROVED** [review]
LGTM🟢

File: ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx [hunk 5]
```diff
@@ -112,15 +113,16 @@ export const components: NotificationComponent<ERC721Notification> = {
...
-            tokenId={notification.data.nft.token_id}
-            tokenName={notification.data.nft.name}
-            tokenSrc={notification.data.nft.image}
-            networkName={nativeCurrencyName}
+            nft={{
+              address: notification.data.nft?.collection.address || '',
+              chainId,
+              image: notification.data.nft?.image || '',
+              name: notification.data.nft?.name || '',
+              standard: TokenStandard.ERC1155,
```
Comment (Prithpal-Sooriya) [inline ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx:123]: Shouldn't we se the standard for this file to ERC721?

If this is blocking the approvals, can we do this in a separate PR?

File: ui/components/multichain/notification-detail-nft/notification-detail-nft.stories.tsx [hunk 6]
```diff
@@ -17,11 +18,13 @@ const Template = (args: NotificationDetailNftProps) => (
...
-  networkSrc:
-    'https://token.api.cx.metamask.io/assets/nativeCurrencyLogos/ethereum.svg',
-  tokenId: 'NFT ID',
-  tokenName: 'NFT Name',
-  networkName: 'Ethereum',
+  nft: {
+    address: '0xAddress',
+    chainId: '0x1',
+    image:
+      'https://i.seadn.io/s/raw/files/a96f90ec8ebf55a2300c66a0c46d6a16.png?w=500&auto=format',
```
Comment (Prithpal-Sooriya) [inline ui/components/multichain/notification-detail-nft/notification-detail-nft.stories.tsx:25]: @MetaMask/notifications when did our notification NFT information start failing to render!

This was before this PR (on develop), investigate what changed.

Screenshot 2024-10-25 at 09 43 35

Comment (Prithpal-Sooriya) [reply ui/components/multichain/notification-detail-nft/notification-detail-nft.stories.tsx:25]: I'm resolving this for our notifications team to investigate.

**Review by Prithpal-Sooriya — APPROVED** [review]

**Review by jclancy93 — APPROVED** [review]

File: ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx (hunk 5 above)
Comment (Prithpal-Sooriya) [reply ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx:123]: In fact, why is the `standard` property added? I checked `nft-item.js` and I don't think it is used. It was a little hard to verify since the file is in JS

File: ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx [hunk 7]
```diff
@@ -112,15 +113,16 @@ export const components: NotificationComponent<ERC721Notification> = {
...
         return (
           <NotificationDetailNft
-            networkSrc={nativeCurrencyLogo}
-            tokenId={notification.data.nft.token_id}
-            tokenName={notification.data.nft.name}
-            tokenSrc={notification.data.nft.image}
-            networkName={nativeCurrencyName}
+            nft={{
+              address: notification.data.nft?.collection.address || '',
+              chainId,
```
Comment (Prithpal-Sooriya) [inline ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx:120]: I've tested this, this does not work correctly.
1. We need to append this chainId with an `0x` at the start.
2. The underlying component  `nft-item.js` only uses the [current network](https://github.com/MetaMask/metamask-extension/blob/4e98d1af3c2b7aadb03335736fe663938075edf6/ui/components/multichain/nft-item/nft-item.js#L35), whereas notifications can come from any network (not just the current network selected)

Comment (Prithpal-Sooriya) [reply ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx:120]: CC @MetaMask/notifications @darkwing 

If this is a PR blocker, I'm fine with ignoring for now if we can get a fix in a diff PR soon (otherwise all notifications will look like they came from the users selected chain)

Comment (matteoscurati) [reply ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx:120]: @Prithpal-Sooriya you are right

Comment (Prithpal-Sooriya) [reply ui/pages/notifications/notification-components/erc721-sent-received/erc721-sent-received.tsx:120]: Should be an easy fix in the `nft-item.js` component.

```js
  const { image, imageOriginal, name, tokenId, chainId } = nft;
  ...
  const currentChain = chainId ?? useSelector(getCurrentNetwork); // use the chainId if it was passed in
```

**Review by infiniteflower — APPROVED** [review]

File: ui/components/multichain/nft-item/nft-item.js [hunk 8]
```diff
@@ -16,27 +16,34 @@ import {
...
-}) => {
+export const NftItem = ({ nft, onClick, clickable, badgeWrapperClassname }) => {
   const testNetworkBackgroundColor = useSelector(getTestNetworkBackgroundColor);
   const isIpfsEnabled = useSelector(getIpfsGateway);
   const openSeaEnabled = useSelector(getOpenSeaEnabled);
 
+  const { image, imageOriginal, name, tokenId } = nft;
+
+  const ipfsGateway = useSelector(getIpfsGateway);
+  const currentChain = useSelector(getCurrentNetwork);
```
Comment (darkwing) [inline ui/components/multichain/nft-item/nft-item.js:35]: This line is problematic; we need a selector that mimics the return value of `getCurrentNetwork` but allows us to pass a chainId instead of assuming the current.
